### ISSUE TITLE
Change chunking algorithm?

### ISSUE THREAD

**Issue by bmschmidt** [massivetexts/htrc-feature-reader#9]
I've made a stab at a modulo-based chunking algorithm in [this commit.](https://github.com/massivetexts/htrc-feature-reader/commit/2925ad93d52e5f6919773661b0ba07b611d6ec5f) I would propose replacing the chunking algorithm with this one, but I'm keeping in a second branch alongside the current version until it's fully compatible with the existing one.

The general strategy is to draw chunk sizes based on precalculated threshold wordcounts, rather than always trying to push up to the target size. I can see three reasons this is better than the current one.

1. It's about 5.5x faster: 100ms vs 550 ms for a typical chunking on my laptop. That's significant time even against the other algorithms we're running, and getting quick enough that we don't need to worry about caching the chunks. Probably there are ways to speed up the current version.
2. It's a simpler to execute in other languages; so we could implement it in the R/javascript packages too. (In fact, more or less I already have, less some improvements I learned from the current version around page length adjustments, etc.)
3. I think we may have been using the wrong minimization function to judge chunking success this summer. We used standard deviation *within* a book. ("does it produce even chunks?"). But really we want to have comparable chunking *across* multiple chunkings of the same book; so we should really be minimizing the standard deviation of chunk size across *all* books, not within one single book. Within the PZ volumes test set, this has a std of chunk size of 381.0
vs 445.9 for the current method.

test code (requires simplechunks branch.)
```python
import pandas as pd
import htrc_features
import glob
import random

vols = [htrc_features.Volume(v) for v in glob.glob("../data/PZ-volumes/*.json.bz2")]
for v in vols:
    print(v)
    v.tokenlist()

l1 = []
l2 = []
for vol in vols:
    l1.append(vol.simple_chunks(10000, 1000).groupby('chunk')['count'].sum())
    l2.append(vol.chunked_tokenlist(10000, 1000).groupby('chunk')['count'].sum())


print(pd.concat(l1).std())

print(pd.concat(l2).std())
```

**Comment by organisciak** [comment]
Cool. I'll take a look!

**Comment by organisciak** [comment]
Incidentally, we should add tests for the chunking to the test suite, but I'm not sure what the measure or 'correctness' should be? Should we just do standard deviation that's less than or equal to what we've seen on the test documents so far?

**Comment by bmschmidt** [comment]
Looking at this a little more carefully, it seems like a lot of this variation is due to the old method offloading incredibly variability to the last chunk in each book. Looking at the std of length in chunks `[:-1]`, the old method overperforms. In practice, there's probably some benefit for books, since the last section is hard to align on in any case. So I may refactor to use your strategy. I can see a way to do it faster/more concisely, I think.

```
First element(s) of list 	Method 0	Std is 528.1367788204653
First element(s) of list 	Method 1	Std is 322.69202478081615
Last element(s) of list 	Method 0	Std is 497.807363774956
Last element(s) of list 	Method 1	Std is 1156.7570720810502
Middle element(s) of list 	Method 0	Std is 316.47870127637646
Middle element(s) of list 	Method 1	Std is 270.7865146574945
```

(Method 0 is the proposed change, method 1 is the current method).

**Comment by bmschmidt** [comment]
Chunking tests can't, I don't think, capture whether the algorithm is good. I think the most important thing is just 

1. That they don't fail in various edge cases where the word count produces 0, 1, or 2 chunks
2. That the constraints are met (no chunks larger than the limits set).

**Comment by organisciak** [comment]
Ah, thanks for bringing that up again! We explicitly made the decision to offload some of the variability to the end, right? Really reaching into my poor memory - we discussed splitting the difference between first and last and decided that just the end should be fine and not require backtracking?

The test may be a good way to identify edge cases. I'll start a separate ticket as a task item.

**Comment by bmschmidt** [comment]
My recollection is that we decided to divide errors between the first and last chunks. But then used a scheme where errors can snowball so the last chunk ended up being worst than the first, anyway, although the first is still worse than the middle ones.

Working on a faster version of the algo you coded, but finding it weirdly difficult! Handling cumulative counts in numpy is a minefield of off-by-one errors.

**Comment by bmschmidt** [comment]
OK, I think I see a mild bug in the current code that explains some discrepancy here. Note this section

```python
            firstchunk_full = (edgeadjust and (chunkname == 1) and (counter > (chunk_target + edgeadjust - avg_page_n/2)))
            regularchunk_full = (counter > (chunk_target - avg_page_n/2))

            
            if (firstchunk_full or regularchunk_full) and len(page_collector) and (chunkname < nchunks):
                  # fold the stuff out.
```

When the first chunk is supposed to be a little longer than a regular chunk, the `regularchunk_full` will always fire before the `firstchunk_full` gets a chance to, so the first chunk differs from the lastchunk despite (I think?) the intent.

No need to fix this if we can retire the code, but just a note on why I'm not trying to target the previous conditions exactly.

**Comment by bmschmidt** [comment]
OK, I've pushed a new version that seems to generally exceed the existing one in [this commit](https://github.com/massivetexts/htrc-feature-reader/commit/b64143e75c026841d51ddad96e25fe8e8c014e59). The new method is called `simple_chunks2`; I'd propose overwriting the existing chunking method with that, and maybe keeping the current simple_chunks method but, er, simplifying it further to not even include front/end adjustments. 

The major differences are:

1. It uses numpy arrays and argmin to handle the loops, so is substantially faster than the current (82ms per volume with the PZ set, vs 448ms for the current method).
2. It looks ahead one slot more, always finding the page break that will make the current chunk closest to the target, even if it goes over by a little bit.
3. It behaves more as intended in balancing overflow across the first and last chunks. (We could switch it to dump all overflow in the last chunk, too.

The parameters are basically the same. Although if our goal is really to offload errors to the first and last chunks, it's not clear to me why we should be using a max_adjust of 1000 as we have been; why not just say max_adjust = 1; in the test set, that reduces the standard deviation of length down to 100 words, which is probably about the best we can do, while keeping the outer chunks to within 7500 to 12,500 words. A couple performance stats:

![image](https://user-images.githubusercontent.com/1110758/71705559-6f1dba00-2dae-11ea-81c4-5ad155cf9b02.png)

And the stats for each algorithm targeting 420 words of variance:

```tsv
	method	chunk_portion	deviation	max	min	time
0	simple_chunks2	First	737.505622	12174	8913	75
1	simple_chunks2	Last	726.758655	12001	9106	75
2	simple_chunks2	Middle	243.611658	10509	9489	75
3	simple_chunks	First	689.430490	11954	8913	78
4	simple_chunks	Last	688.223019	12001	9106	78
5	simple_chunks	Middle	266.303845	10598	9420	78
6	chunked_tokenlist	First	275.211866	10412	9454	483
7	chunked_tokenlist	Last	2029.362673	14572	6374	483
8	chunked_tokenlist	Middle	184.024218	10399	9650	483
```

**Comment by organisciak** [comment]
This sounds good to me, I think we can replace the original chunked_tokenlist. I'm not sure I gathered why you recommend keeping simple_chunks alongside the new one? And if we keep both, what are your thoughts on organizing the code by having them both as private functions, where chunked_tokenlist loads the tokenlist and, based on a `strategy=''` argument, sends it to one or the other? In addition to neatness, it keeps open the possibility to call the chunking method outside of the Volume.

I'm realizing tests should focus on cases more than quality - making sure that the code will work with all the arg possibilities. I _really_ need to write those, just backlogged from an extended cold.

**Comment by bmschmidt** [comment]
Here's what I'd propose, in docstring format.

It drops two args for simplicity:

1. "max_adjust". As discussed above, setting it to 1 means that we get evenly spaced middle chunks, and end chunks are only off by > 25% in hopeless situations (e.g., the document is less than half the length of the minimum chunk size. In practice, I have no intuition about how to set this well.
2. "suppress_warnings." More controversial, but: can't we just keep `self._tokencounts` from ever having chunk information in it? If we want to attach methods for caching to the `self` or to pandas, that's great, but it just seems unsafe to have _tokencounts returning things like this to me. And we're now down under a tenth of a second per volume, which seems good enough to me to not even cache.

```python
    def chunked_tokenlist(self, chunk_target = 10000, overflow_strategy = "ends", page_ref=False, **kwargs):
        '''
        
        Return a tokenlist dataframe grouped by numbered 'chunks', each of which has roughly `chunk_target` words.

        chunk_target: the target size--in number of words--of each chunk.

        overflow_strategy: How to handle cases in which the total number of words does not divide directly into the chosen chunk_target.
        - "ends" allows the first and last chunks -- which, in books, are often the messiest -- to vary greatly in length
          while keeping the middle sections as close to `chunk_target` in length as possible given the page lengths.
        - "even" sets preset targets based on the document length, and creates chunks that are of approximately even size within each book.
          There may be great variability in chunk length *between* books using this method.
        - "last" (not implemented) keeps all but the last chunk at the desired length, but allows huge variability in the final chunk.
        
        page_ref: Whether to include columns in the frame identifying the beginning and ending page for each chunk.

        '''
```

**Comment by bmschmidt** [comment]
I think we agree here, and I'll close the issue when I can check it in.

**Comment by organisciak** [comment]
Coming back to this. I think the default (or only) overflow_strategy should be 'last', as we opted earlier. The start of a book certainly *can* vary greatly in some books, but the "ends" strategy would greatly increase the number of books with misaligned chunks from the start.

**Comment by bmschmidt** [comment]
Hmm. I don't think we did opt 'last' earlier--we also had the first chunk [varying in length](https://github.com/massivetexts/htrc-feature-reader/commit/62879babcf7b5d927ba79760eba9bc3284ec7cc3#diff-c91e09150d227c05eebbde960b0bd309L1030).

In both, last suffers more than first because the error can accumulate as earlier chunks are lined up, though. 

I guess my issue here is that I think first and last will still be usable at 7,500 or 12,500 words compared to 10K, but that a last chunk that varies between 1 word and 9,999 words is kinda useless.

I also slightly worry that trying to engineer perfect end-first alignment is kind of a trick. My mind keeps going sort of the opposite direction, which is that we should have two or three chunkings of each book with different offsets in the index.

Anyway, I don't feel strongly about this. Happy to defer.

**Comment by organisciak** [comment]
Yeah, I noticed last night that I'd overlooked your message that "My recollection is that we decided to divide errors between the first and last chunks." Sorry! Hopefully I didn't sound bullheaded, that was unintentional.

I love your code. Very elegant. The main loop is nice!

I've been running it on larger random batches and adjusted as outliers come up. One thing that threw me off that I couldn't get an 'even' strategy to work - it turned out that it was because but "middle of the book" stats had a much larger N (a book on has one first, one end, but many middle), so they always look better because it starts to converge toward 0. Only sampling the size of one middle chunk per book was more accurate.

I'm fine with 'ends' - testing this made me realize that the general differences are small. Also, a proper 'last' strategy would require re-calculating overflow each time you set a breakpoint, I think (I only tested that with a few hundred texts before retreating from the messing code).

Here are the stats for 30k books using the currently checked-in strategies:

```
========Strategy 'ends' with target of 10000=======
First	| Difference From Target: 233 	| Std: 1959
Middle	| Difference From Target: 217 	| Std: 1598
End	| Difference From Target: 173 	| Std: 3301

========Strategy 'even' with target of 10000=======
First	| Difference From Target: 219 	| Std: 1511
Middle	| Difference From Target: 219 	| Std: 1526
End	| Difference From Target: 163 	| Std: 3030

========Strategy 'last' with target of 10000=======
First	| Difference From Target: 205 	| Std: 1420
Middle	| Difference From Target: 220 	| Std: 1764
End	| Difference From Target: 199 	| Std: 4059
```

Mean differences look good, but those stds look high? I'll check after my 5hrs (!!) of meetings are done.

**Comment by bmschmidt** [comment]
Those stds are too high, yeah. Something must be wrong. Will take a look if I get a chance, too.

I can't shake the idea that the best 'even' strategy probably works out by setting the middle partition first, and then recursively working out towards the ends, rather than letting errors accumulate toward the tail. But I tried this over the summer and never got it to work.

I was thinking simply to do what I do in R and Javascript, which is:
1. Determine how chunks
2. Set breakpoints every floor((n_words + 1)/n_chunks) words.
3. Start a new chunk at every point the cumsum is above those lines.

The results are a little different--it treats 10,000 words as an absolute ceiling, not a target--but it's easy to implement in different languages and straightforward to understand.

**Comment by bmschmidt** [comment]
Oh yeah, and I hope we've both realized by now that it's impossible to register concerns about code at this level without coming off like an insane person. No worries about nitpicking or forgetting.

Although--seriously--you should maybe think about adding a code of conduct to the main branch.

**Comment by organisciak** [comment]
How are the distributions of sizes with the ceiling approach?

I should have kept book-specific data, so re-running now. I saw a few wacky outliers that I'd like to identify, like `last_chunks` that are waaay over. Maybe I introduced some bug?

**Comment by organisciak** [comment]
Found an outlier. It's really long books with longer word counts where the average chunk size is undersized - if there are 81 chunks averaging 9.6k, the 82nd chunk will have >40k words. I assume that books on the other end, where the average length is higher than the target, were failing and being caught by exception handling.

We can tackle this multiple ways. One is to allow our nchunks to grow if necessary. The other is to adjust targets dynamically. I'll try an idea for the latter.

**Comment by organisciak** [comment]
I checked in a proposed correction, currently in 42a424b, which on first blush looks good. The other issue contributing to the large `std` is 1 and 2 chunk books. Not much we can do about those so excluding those from diagnostics.

Running another large run now, and I'll examine distributions and outliers later.

**Comment by organisciak** [comment]
I should explain the strategy - it simply re-calculates the remaining words per chunk and adjusts the target for the next chunk. So if multiple chunks have been undersized, the next one will target something bigger.

For the 'even' strategy it tries to adjust based on how much it's off its target, while for the other strategies it has a decay, so it becomes less aggressive as it goes on. This counterbalances the fact that adjustments are less distributed toward the back; e.g. being 10k words off with 50 chunks left is different than having three chunks left, so in the latter case only adjust a small proportion of the 10k in [-3:-2], leaving most of the weirdness for [-1].

I can see this requiring a 'max adjust' again, though referring to the maximum adjustment for middle chunks rather than start/end chunks like before, and probably hardcoded because the scenarios where it would be needed are limited. If a book starts with some odd, super-long pages, I'd rather sacrifice the end for *some* usable chunks.

**Comment by bmschmidt** [comment]
Ah, that makes sense as a failure case. But given these choices:

> 1. One is to allow our nchunks to grow if necessary. 
> 2. The other is to adjust targets dynamically. I'll try an idea for the latter.

I think the former (reallocating nchunks as necessary) is better. The goal is simply to create as many 10,000 word chunks as possible. We don't get anything good out of having 20 10,500 word chunks instead of 21 10,000 chunks.

**Comment by bmschmidt** [comment]
But wait. The explanation you gave explains high std in the final chunks. But how does it explain it in the middle chunks? Is it there are a bunch of 0-sized middle chunks caused by this same bug?

```
Middle	| Difference From Target: 219 	| Std: 1526
```

**Comment by bmschmidt** [comment]
I think what this is all suggesting to me is that it's really hard to avoid either.

1. Clobbering the last. (or the first, or the middle, but we've sensibly chosen the last as the least useful.) Trying to use offset on the first the keep the last from getting way off is like aiming a fire hose from ten feet back. It's possible that an 'ends-in' approach would work.
2. Creeping errors where some books return consistently too large. I keep turning in my head to the idea that 'even' should be a recursive partition strategy. But that, like the overflow allocation, ends up leading to contiguous chunks all erring in the same direction. I suspect this might be bad for matching especially, because it means that two copies of the same book might have a series of 9,000 word chunks in the critical sections of one, and 11,000 word chunks in the critical sections of the other, again and again. That seems like it might be bad.

One option that might work on extremely long works is an 'ends-in' approach where we merge together a 'clobber first' and 'clobber last' approach.  May give that a shot. But given what you've said, I do think a lot of these will simply be blocked by a few massive pages.

So I'm starting to think we should just bite the bullet and do 'clobber last'. The question there is whether a rump last section of (say) 100 words should be a miniscule chunk, or should be added on to the prior chunk to make it a 'bad last' of 10100 words.

**Comment by organisciak** [comment]
The middle std was off because for 1-chunk books, that same chunk would be counted as first/middle/last.

**Comment by organisciak** [comment]
Here are the stats for 50k books in each strategy:

![image](https://user-images.githubusercontent.com/58351/72946525-70eaf580-3d3c-11ea-8388-171fc69d6b97.png)

The new adjustment works fine, at least for 'even' and 'last'. I'll still explore outliers from that run. I think I'm coming back to the strategy of the original code, to mix both methods - distributing some of the offset across all pages, while allowing for the end to take the brunt of the force.

> The goal is simply to create as many 10,000 word chunks as possible. We don't get anything good out of having 20 10,500 word chunks instead of 21 10,000 chunks.

To be clear, the adjustment just encourages the pendulum to swing the other way. If you have a 10.5k word chunk, it's trying to *prevent* the next one from also being 10.5k. It will target 9500-9999 words for the next chunk, hopefully encouraging a cutoff a page earlier. That encourages the sizes to hang out around 10k within a book.

Assuming the outliers from the last run don't give us new things to think about, I plan to update the code to have a cap on the adjustment, *and* allow new extra chunks in the cases where it's still chronically undersized.

**Comment by organisciak** [comment]
Here's the distribution of deviations from target, by the way:

![image](https://user-images.githubusercontent.com/58351/72947566-a2b18b80-3d3f-11ea-87b5-afbcf6b4840e.png)

**Comment by bmschmidt** [comment]
My fear is that if you have a 12,000 word chunk (because of one weird page, or something) 
the next 4 chunks will all be 9,500 to compensate. That sort of local auto-correlation makes me nervous because I can imagine it subtly making matching harder on individual books in ways that are tricky to debug. The books that end up with lots of 9,500-word chunks may have a variety of features different than the ones that end up with lots of 10,500 chunks.

In the checked-in code, the 'ends' strategy is giving me a lot of results like this: (mdp.39015013407237)

```
1    7655
2    9453
3    9626
4    9576
5    9834
6    9496
7    9515
```

Where the first chunk is real small, but then the last chunk actually does just as well as the middle ones. And all of the volumes are skewed in the same direction.

---

Separately, I think the following line in the check-in code needs an absolute value:
```python
adjust = min(max_adjust, adjust)
```

**Comment by bmschmidt** [comment]
Putting this more concretely. Two indicators of a good algorithm worth tracking.

In addition to mean chunk length would be means aggregated across books. In addition to wanting all our chunks to cluster around 10,000, we want each book to cluster around 10,000 words.

A further extension is that we want each pair adjacent chunks in each book to cluster around 10,000 words.

```python

all_book_averages = []
adjacent_chunk_averages = []
for book in books:
   mid_chunks = chunk_counts[1:-1]
   if (len(mid_chunks) > 0):
       all_book_averages.append(np.mean(mid_chunks))
       for i in range(len(mid_chunks)):
             # (needs error handling)
             adjacent_chunk_averages.append(np.mean(mid_chunks[i:i+2]))

```

Then we want to minimize the spread of all_book_averages and adjacent_chunk_averages.

**Comment by organisciak** [comment]
Thanks for catching that bug. I also changed the adjustment so that it ignores the final chunk (unless seeking `even`).

Your example volume made be realize that "ends" is at odds with the n_chunk adjustment, since you punish the first chunk on a guess on how many chunks there will be, but if the offset keeps just on one side of chunk_target, a new chunk will be added or one will be removed, and the last chunk won't be punished in the same way as the first. Should we turn off the auto chunk adjustment for the 'ends' policy? I think its fine.

**Comment by bmschmidt** [comment]
Yeah that's just what I mean by holding a firehose. Basically, you have to do a lot of damage to the first chunk and it *still* doesn't keep the last chunk from being even more ragged, because you can't really control the direction the overflow moves in. The problem is independent of the n_chunks adjustment. This is why I think that maybe we should just adopt a 'clobber the last strategy'; all other chunks will get the very tight corner from the lower left of your histogram set, and the last one will just be a mess.

Another option for 'ends' would be:

1. Build the front and last elements *initially* to both be about 8,000 words (or whatever the 'smaller' size is, removing them from the document.
2. Proceed with the 'even' chunking strategy for all the interior chunks, including the overflow offset targets you suggest, but (IMO) without any 'max_adjust' factor.

Another option that has some merit is to work in from both ends simultaneously. That is, apply 'clobber ends' again and again, but with something very close to 10K for every pair but the first. I might try to implement this quickly.

**Comment by bmschmidt** [comment]
Hah, 'quickly.'

Here's a self-contained function that moves forward from both ends towards the middle. Clobber ends, repeatedly. Since it does two chunks at a time, there's much less need for any max_overflow argument; and it mimics the general reason that we're doing end clobbering to begin with. Haven't fully checked for off-by-one errors.

OVERALL STANDARD DEVIATIONS
9648.616996507568 2049.685985487963
STANDARD DEVIATIONS EXCLUDING ENDS
10000.941686300908 293.77731610835593


```python
def end_chunks(page_counts, target):
    
    # Register front and back offsets
    position = [0, len(page_counts)]

    breaks = np.zeros(page_counts.shape[0])
    breaks[0] = 1
    
    while True:
 #       if page_counts.shape[0] == 0:
 #           break
        forward = np.cumsum(page_counts)
        backward = np.cumsum(np.flip(page_counts))
        words_left = forward[-1]
        #What is this number supposed to be?
        if words_left < (target * 1.5):
            break

        overflow = words_left % target
        if (target - overflow) < overflow:
            overflow = -(target - overflow)
        # Split the overflow across the ends
        loc_target = target + overflow/2
        #What is this number supposed to be?    
        if words_left < (target * 2.5):
            midpoint = np.argmin(np.abs(forward - words_left/2))
            breaks[midpoint + position[0] + 1]  = 1
            break

        best_front = np.argmin(np.abs(forward - loc_target))
        best_back = np.argmin(np.abs(backward - loc_target))
#        print(forward[best_front] - loc_target, backward[best_back] - loc_target)

        position[0] = position[0] + best_front + 1
        # TODO is there an off-by-one indexing error on either front or back below?
        position[1] = position[1] - best_back - 1
        breaks[position[0]] = 1
        breaks[position[1]] = 1

        page_counts = page_counts[(best_front + 1):(page_counts.shape[0] - best_back - 1)]
        
    return np.cumsum(breaks)
```

**Comment by organisciak** [comment]
I thought the 'max_adjust' was in response to your (valid!) concern about swaying the middle chunks too far.

I really don't want to keep introducing complexity, but I keep thinking about how we clobber the last chunk. If there's 13k words left, do we have one chunk of 13k, or one of 10k then one more of 3k? The former looks better with the statistics that I've been running, but the latter is *probably* what we want? I'm coming around to a negative skew on the size of the last chunk.

I don't have time to test your code today or tomorrow, but generally, here are the rules that I'm roughly looking for, where the 'diff' values are the size of the chunk minus that target size:

- 'ends'
    - lastdiff-firstdiff should have a mean close to zero and low std
    - middiff should have a zero mean and low std
    - min/max for lastdiff and first diff (when run on a test set) should be within 1/4 of chunk target (else there may have been a need for a different chunk count)

- 'even'
   - means and stds should be low and similar in all categories, min/max should be within 1/4 of chunk target

- 'last'
   - middiff should have a zero mean and low std, min/max should be low
   - firstdiff should be similar to middiff
   - lastdiff - ignore std, negative mean is good (large final chunks should be split into one good and one throwaway), min can be small, max should be conservative (don't go greatly over before starting a new chunk)

**Comment by bmschmidt** [comment]
Here's a notebook with some evaluations. It can run without any of the other feature reader code, and we can just patch those in when we decide.

https://github.com/massivetexts/htrc-feature-reader/blob/4598652722a174f8ebf60e7d7a54d1b3b3d095de/examples/Test%20chunking.ipynb

My takeaway is that a two-ended strategy makes a 'clobber the ends, but not too much' strategy tractable.

And also that in addition to your metrics, we really should consider minimizing intrabook error. Iit's bad to have a book all at 9500 words in the mid chunks and another one hovering around 10,500, because they might not match as well.

**Comment by bmschmidt** [comment]
I've added 55,000 randomly sampled page counts to the file linked at the head of that notebook (not citing here because of crawlers) instead of the 5K used in the example. Slowly rerunning the stats and batching together.

**Comment by bmschmidt** [comment]
Some evaluations.

Note: I also corrected some code that was causing an infinite loop on all methods where the last page had 27,000 (!!!) words, so it's important that we refactor back from the notebook into the mainline.

The first is the same metrics you were using. I take it to mean that either 'two sided clobber ends', 'clobber the last,' 'current even' are good, but that 'two-sided clobber ends then even' has extremely good performance on the middle chunks--as good as 'clobber the last'--without any completely junk final chunks. (i.e., none less than 6,300 words. Aforementioned 27,000-word final chunk makes the 'max' useless here).

The second addresses my concerns about bookwise skew. This is much better on the two-sided approaches than the current one, which frequently has books with middle chunks averaging 400 words off of the target. The 'ends then even' strategy is only a tiny bit worse.

So my proposal is to keep 'even,' 'ends', and 'last' using the two-sided formulation for the first two, using the 'ends then even' strategy for 'ends' (rather than continuing to apply 'ends'  recursively.) It might make sense to bundle 'dumb' as well.

For our project, I like a default of 'ends', but am open to a default of 'last'.

All method have a comparable runtime (4.7 ms per book--wow that's fast!), except for 'dumb', which is ~8x faster.

## chunk-level variations

![image](https://user-images.githubusercontent.com/1110758/73088391-208da800-3ea2-11ea-832f-46b0cf244fdc.png)

## Variations in the aggregated difference in books

i.e.; does this book have an average middle chunk length of 9,700 instead of 10,000? because it shouldn't. 

![image](https://user-images.githubusercontent.com/1110758/73088419-313e1e00-3ea2-11ea-9230-bf4a1e836962.png)

**Comment by bmschmidt** [comment]
Two things remaining here.

1. For me to write a couple tests that act on the Volume class; currently tests only deal with the underlying algorithm. I'm not sure what happens in some limit cases, especially when working with a volume that has no pages.
2. Some indexing issues. The values returned from this algorithm don't have their indexes set right.

**Comment by bmschmidt** [comment]
Indexing is fixed, chunking is now tested on the volume in https://github.com/massivetexts/htrc-feature-reader/pulls
